[gpu] Use clustered gpu.subgroup_reduce for nested layout distribution

andfau-amd opened the pull request:
There is now support in MLIR for expressing a subgroup reduction operation that operates on several "clusters" in parallel, so it is no longer necessary to build a series of shuffles.

It has been verified that, at least if the upstream patterns are used, the resulting sequence of shuffles is the same as the old code.

This commit also adds a new pass, ExpandGPUOps, which uses the upstream patterns to expand these ops, and adds it to the LLVMGPU pass list.

Resolves #18142.

Groverkss commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp:
```diff
@@ -513,8 +514,16 @@ struct DistributeMultiReduction final
         if (reductionMask[i]) {
-          extracted = doPackedThreadReductionOnDim(rewriter, layout, extracted,
-                                                   kind, i);
+          Location loc2 = extracted.getLoc();
```
You can just use `loc`. `extracted` has the same location assigned before.

andfau-amd replied:
Oh you're right, not sure how that came about. What's the purpose of the locations system anyway?

Groverkss reviewed (approved):
LGTM! There seems to be a failure in some tests because of this patch. Looks like you forgot to add a lowering for this op to gpu.shuffle ops?

```
failed to legalize operation 'gpu.subgroup_reduce' that was explicitly marked illegal
```

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp:
```diff
@@ -513,8 +514,15 @@ struct DistributeMultiReduction final
-                                                   kind, i);
+          int64_t offset = getShuffleOffset(layout, i);
+          int64_t width = getShuffleWidth(layout, i);
+          if (offset > UINT32_MAX || width > UINT32_MAX)
```
Prefer `std::numeric_limits`

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp:
```diff
@@ -513,8 +514,15 @@ struct DistributeMultiReduction final
       // Reduce across all reduction dimensions 1-by-1.
       for (unsigned i = 0; i < reductionMask.size(); ++i) {
```
```suggestion
      for (unsigned i = 0, e = reductionMash.size(); i != e; ++i) {
```

andfau-amd replied:
I didn't write this part and I think readability would suffer for this.

kuhar replied:
no, this is the standard way to write loops in the llvm coding standards: https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

kuhar replied:
if anything, readability improves because you communicate that the size of the list doesn't change throughout the loop

qedawkins commented on .gitmodules:
```diff
@@ -3,7 +3,7 @@
 [submodule "third_party/llvm-project"]
 	path = third_party/llvm-project
-	url = https://github.com/iree-org/llvm-project.git
+	url = https://github.com/andfau-amd/llvm-project.git
```
(reminder to drop this before landing)

qedawkins commented on compiler/src/iree/compiler/Codegen/Common/GPU/ExpandGPUOps.cpp:
```diff
@@ -0,0 +1,50 @@
+    mlir::populateGpuBreakDownSubgroupReducePatterns(
+        patterns, /* maxShuffleBitwidth=*/32, PatternBenefit(2));
+    mlir::populateGpuLowerClusteredSubgroupReduceToShufflePatterns(
+        patterns, *subgroupSize);
```
mini-nit: If this can take a benefit too, would be good to add so that the order of patterns is explicit.

qedawkins reviewed (approved):
Looks good to land once integrate pulls in the needed commits.

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/ExpandGPUOps.cpp:
```diff
@@ -0,0 +1,50 @@
+namespace {
+
+struct ExpandGPUOpsPass final : impl::ExpandGPUOpsPassBase<ExpandGPUOpsPass> {
+  ExpandGPUOpsPass() {}
```
Do we need this?

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp:
```diff
@@ -511,10 +512,18 @@ struct DistributeMultiReduction final
 
       // Reduce across all reduction dimensions 1-by-1.
-      for (unsigned i = 0; i < reductionMask.size(); ++i) {
+      for (unsigned i = 0, n = reductionMask.size(); i < n; ++i) {
```
nit: the idiomatic form would be
```suggestion
      for (unsigned i = 0, e = reductionMask.size(); i != e; ++i) {
```

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp:
```diff
@@ -511,10 +512,18 @@ struct DistributeMultiReduction final
+          int64_t width = getShuffleWidth(layout, i);
+          if (offset > std::numeric_limits<uint32_t>::max() ||
+              width > std::numeric_limits<uint32_t>::max())
+            return failure();
```
I think this can be an assertion instead?

kuhar commented on compiler/src/iree/compiler/Codegen/Utils/GPUUtils.h:
```diff
@@ -156,6 +157,11 @@ Value unpackToVector(Location loc, OpBuilder &builder, Value packedInput,
 Value getCombiningIdentityValue(Location loc, OpBuilder &builder,
                                 vector::CombiningKind kind, Type identityType);
+
+/// Returns the matching GPU reduction operation.
+mlir::gpu::AllReduceOperation
+combiningKindToAllReduce(vector::CombiningKind kind);
```
Is this used anywhere?

andfau-amd replied:
Yes, the single use is introduced by this commit.

kuhar replied:
Oh, I've just found it, never mind

kuhar reviewed (commented):
Should we have a basic LIT test to check that the subgroup size gets correctly passed on to the subgroup expansion patterns?

kuhar commented on compiler/src/iree/compiler/Codegen/Common/GPU/ExpandGPUOps.cpp:
```diff
@@ -0,0 +1,48 @@
+    RewritePatternSet patterns(ctx);
+    mlir::populateGpuBreakDownSubgroupReducePatterns(
+        patterns, /* maxShuffleBitwidth=*/32, PatternBenefit(2));
+    mlir::populateGpuLowerClusteredSubgroupReduceToShufflePatterns(
```
nit: we don't need mlir since this code is already in the mlir namespace
```suggestion
    populateGpuBreakDownSubgroupReducePatterns(
        patterns, /* maxShuffleBitwidth=*/32, PatternBenefit(2));
    populateGpuLowerClusteredSubgroupReduceToShufflePatterns(
```

kuhar reviewed (commented):
looks good, let's wait for the mlir commit to propagate to iree

andfau-amd committed:
[gpu] Use clustered gpu.subgroup_reduce for nested layout distribution

There is now support in MLIR for expressing a subgroup reduction
operation that operates on several "clusters" in parallel, so it is no
longer necessary to build a series of shuffles.

It has been verified that, at least if the upstream patterns are used,
the resulting sequence of shuffles is the same as the old code.

This commit also adds a new pass, ExpandGPUOps, which uses the upstream
patterns to expand these ops, and adds it to the LLVMGPU pass list.

Resolves #18142.

Signed-off-by: Andrea Faulds <andrea.faulds@amd.com>

kuhar reviewed (approved):
LGTM
